include/RED4ext/HashMap.hpp

```diff
@@ -4,89 +4,126 @@
+                {
+                    break;
+                }
+
+                while (idx != INVALID_INDEX)
+                {
+                    Node* node = reinterpret_cast<Node*>(nodes + idx * stride);
+                    if (!node)
+                    {
+                        break;
+                    }
+
+                    aFunctor(node->key, node->value);
+
+                    idx = node->next;
+                }
             }
+        }
 
+        T* Get(const K& aKey) const
+        {
+            Node* found = nullptr;
+            uint32_t hashedKey = Hasher{}(aKey);
+            uint32_t idx = indexTable[hashedKey % capacity];
             while (idx != INVALID_INDEX)
             {
                 Node* node = reinterpret_cast<Node*>(nodes + idx * stride);
-                if (!node)
+                if (node->index == hashedKey && node->key == aKey)
                 {
+                    found = node;
                     break;
                 }
 
-                aFunctor(node->key, node->value);
-
                 idx = node->next;
             }
-        }
-    }
 
-    // TODO: Support other hashing methods, this assumes FNV1a64, GameInstance uses CRC32
-    T* Get(const K aKey) const
-    {
-        Node* found = nullptr;
-        uint32_t idx = indexTable[static_cast<uint32_t>(aKey) % capacity];
-        while (idx != INVALID_INDEX)
-        {
-            Node* node = reinterpret_cast<Node*>(nodes + idx * stride);
-            if (node->index == static_cast<uint32_t>(aKey) && node->key == aKey)
+            if (found)
             {
-                found = node;
-                break;
+                return &found->value;
             }
 
-            idx = node->next;
+            return nullptr;
         }
 
-        if (found)
+        IMemoryAllocator* GetAllocator() const
         {
-            return &found->value;
+            return reinterpret_cast<IMemoryAllocator*>(&allocator);
         }
 
-        return nullptr;
-    }
-
-    uint32_t* indexTable; // 00
-    uint32_t size;        // 08
-    uint32_t capacity;    // 0C
-    uintptr_t nodes;      // 10
-    uint32_t unk20;       // 18
-    uint32_t stride;      // 1C - sizeof(Node)
-    uint32_t unk28;       // 20
-    uint32_t unk2C;       // 24
-};
-RED4EXT_ASSERT_SIZE(RED4EXT_ASSERT_ESCAPE(HashMapBase<uint64_t, void*>), 0x28);
-
-template<typename K, typename T>
-struct HashMap : HashMapBase<K, T>
-{
-    virtual ~HashMap(){}; // Empty impl to indicate there's a vtable
-};
-RED4EXT_ASSERT_SIZE(RED4EXT_ASSERT_ESCAPE(HashMap<uint64_t, void*>), 0x30);
+        uint32_t* indexTable;   // 00
+        uint32_t size;          // 08
+        uint32_t capacity;      // 0C
+        uintptr_t nodes;        // 10
+        uint32_t unk18;         // 18
+        uint32_t stride;        // 1C - sizeof(Node)
+        uint32_t unk20;         // 20
+        uint32_t unk24;         // 24
+        uintptr_t allocator;    // 28
```

Comment:
Probably safe to just use IMemoryAllocator* here if we're sure that's what it is, no need to cast it in the getter.

Reply:
it's not a pointer, i'm casting &allocator into a pointer because IMemoryAllocator contains pure virtual funcs
it should be "IMemoryAllocator allocator"

Reply:
Oh it's actually the vtable at the bottom of the class I see. Could make a stub version with empty impl and place it there but I'm not sure that's worth the effort.